synapseclient/api/entity_factory.py

```diff
@@ -0,0 +1,391 @@
+
+    bundle = await get_entity_id_version_bundle2(
+        entity_id=entity_id,
+        version=entity_version,
+        synapse_client=synapse_client,
+    )
+    syn.cache.add(
+        file_handle_id=bundle["entity"]["dataFileHandleId"], path=filepath, md5=md5
+    )
+
+    return bundle
+
+
+async def _cast_into_class_type(
+    entity_bundle: Dict[str, Any],
+    download_file: bool = True,
+    download_location: str = None,
+    if_collision: str = None,
+    submission: str = None,
+    follow_link: bool = False,
+    entity_to_update: Union["Project", "File", "Folder"] = None,
+    *,
+    synapse_client: Optional["Synapse"] = None,
+) -> Union["Project", "File", "Folder"]:
+    """
+    Take an entity_bundle returned from the Synapse API and cast it into the appropriate
+    class type. This will also download the file if `download_file` is set to True.
+    Additionally, if `entity_to_update` is passed in, the fields of the existing entity
+    will be updated instead of creating a new instance. If the entity is a link and
+    `follow_link` is set to True, the target entity will be retrieved.
+
+    Arguments:
+        entity_bundle: The entity bundle to cast into a class type.
+        download_file: Whether associated files should be downloaded.
+        download_location: The directory to download the file to.
+        if_collision: Determines how to handle file collisions. May be:
+
+                    - `overwrite.local`
+                    - `keep.local`
+                    - `keep.both`
+        submission: The ID of the submission to which the entity belongs.
+        follow_link: Whether to follow a link to its target. This will only do a single
+            hop to the target of the link.
+        entity_to_update: An existing entity class instance to update with data from
+            Synapse.
+        synapse_client: If not passed in or None this will use the last client from
+            the `.login()` method.
+
+    Returns:
+        A Synapse entity object.
+
+    Raises:
+        ValueError: If the entity type is not supported.
+    """
+    from synapseclient import Synapse
+
+    syn = Synapse.get_client(synapse_client=synapse_client)
+
+    # If Link, get target ID entity bundle
+    if (
+        entity_bundle["entity"]["concreteType"] == concrete_types.LINK_ENTITY
+        and follow_link
+    ):
+        target_id = entity_bundle["entity"]["linksTo"]["targetId"]
+        target_version = entity_bundle["entity"]["linksTo"].get("targetVersionNumber")
+        entity_bundle = await get_entity_id_version_bundle2(
+            entity_id=target_id, version=target_version, synapse_client=synapse_client
+        )
+    entity = entity_bundle["entity"]
+
+    from synapseclient.models import Annotations
+
+    annotations = Annotations.from_dict(
+        synapse_annotations=entity_bundle.get("annotations", None)
+    )
+
+    from synapseclient.models import File, Folder, Project
```

Comment:
I think you already import this up top?

Reply:
I am - however it's behind a `TYPE_CHECKING` if statement:
```

if TYPE_CHECKING:
    from synapseclient import Synapse
    from synapseclient.models import File, Folder, Project
```

What it means in this context is that I can reference the type in an argument of a function like:
```
def my_function(file: "File") -> None:
```

However I cannot create an instance of a File like this, the following is NOT valid:
```
def my_function(file: "File") -> None:
    instance_of_file = File()
```
![image](https://github.com/Sage-Bionetworks/synapsePythonClient/assets/17128019/581a67a5-6444-49f7-a342-085c5d0887cc)


This is another instance of lazy importing to prevent circular dependency trees. Because `File, Folder, Project` themselves are importing this factory function to be called from within their respective `get_async()` function calls.